pkg/serviceaccounts/controllers/docker_registry_service.go

```diff
@@ -0,0 +1,213 @@
+
+	RegistryServiceName string
+
+	DockercfgController *DockercfgController
+}
+
+// NewDockerRegistryServiceController returns a new *DockerRegistryServiceController.
+func NewDockerRegistryServiceController(cl client.Interface, options DockerRegistryServiceControllerOptions) *DockerRegistryServiceController {
+	e := &DockerRegistryServiceController{
+		client: cl,
+	}
+
+	_, e.serviceController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return e.client.Services(options.RegistryNamespace).List(labels.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return e.client.Services(options.RegistryNamespace).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Service{},
+		options.Resync,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    e.serviceAdded,
+			UpdateFunc: e.serviceUpdated,
+			DeleteFunc: e.serviceDeleted,
+		},
+	)
+	e.registryServiceName = options.RegistryServiceName
+	e.dockercfgController = options.DockercfgController
+
+	return e
+}
+
+// DockerRegistryServiceController manages ServiceToken secrets for Service objects
+type DockerRegistryServiceController struct {
+	stopChan chan struct{}
+
+	client client.Interface
+
+	registryServiceName string
+
+	dockercfgController *DockercfgController
+
+	serviceController *framework.Controller
+}
+
+// Runs controller loops and returns immediately
+func (e *DockerRegistryServiceController) Run() {
+	if e.stopChan == nil {
+		e.stopChan = make(chan struct{})
+		go e.serviceController.Run(e.stopChan)
+	}
+}
+
+// Stop gracefully shuts down this controller
+func (e *DockerRegistryServiceController) Stop() {
+	if e.stopChan != nil {
+		close(e.stopChan)
+		e.stopChan = nil
+	}
+}
+
+func getServiceLocation(service *api.Service) string {
+	hasPortalIP := (len(service.Spec.PortalIP) > 0) && (net.ParseIP(service.Spec.PortalIP) != nil)
+	if hasPortalIP && len(service.Spec.Ports) > 0 {
+		return fmt.Sprintf("%s:%d", service.Spec.PortalIP, service.Spec.Ports[0].Port)
+	}
+
+	return OpenshiftDockerURL
+}
+
+// serviceAdded reacts to the creation of a docker-registry service by updating all service account dockercfg secrets and
+// changing all interestedURLs
+func (e *DockerRegistryServiceController) serviceAdded(obj interface{}) {
+	service := obj.(*api.Service)
+	if service.Name != e.registryServiceName {
+		return
+	}
+
+	if err := e.handleLocationChange(getServiceLocation(service)); err != nil {
+		glog.Error(err)
+	}
+}
+
+// serviceUpdated reacts to the creation of a docker-registry service by updating all service account dockercfg secrets and
+// changing all interestedURLs, if needed
+func (e *DockerRegistryServiceController) serviceUpdated(oldObj interface{}, newObj interface{}) {
+	oldService := oldObj.(*api.Service)
+	newService := newObj.(*api.Service)
+	if newService.Name != e.registryServiceName {
+		return
+	}
+	if oldService.Spec.PortalIP == newService.Spec.PortalIP {
```

Comment:
compare `getServiceLocation(oldService)` and `getServiceLocation(newService)` to pick up port changes